File: sdk/go/auditlogs.go
package qeetid

import (
	"context"
	"net/http"
	"net/url"
	"strconv"
)

type AuditLog struct {
	ID           string         `json:"id"`
	TenantID     string         `json:"tenant_id"`
	ActorID      string         `json:"actor_id,omitempty"`
	ActorType    string         `json:"actor_type,omitempty"`
	Event        string         `json:"event"`
	ResourceType string         `json:"resource_type,omitempty"`
	ResourceID   string         `json:"resource_id,omitempty"`
	IPAddress    string         `json:"ip_address,omitempty"`
	UserAgent    string         `json:"user_agent,omitempty"`
	Metadata     map[string]any `json:"metadata,omitempty"`
	Hash         string         `json:"hash,omitempty"`
	CreatedAt    string         `json:"created_at"`
}

type AuditLogListParams struct {
	Event   string
	ActorID string
	From    string
	To      string
	Limit   int
	Cursor  string
}

type AuditLogPage struct {
	Data       []AuditLog
	NextCursor string
}

type AuditLogs struct{ http *httpClient }

func (r *AuditLogs) List(ctx context.Context, tenantID string, params AuditLogListParams) (*AuditLogPage, error) {
	q := url.Values{}
	if params.Event != "" {
		q.Set("event", params.Event)
	}
	if params.ActorID != "" {
		q.Set("actor_id", params.ActorID)
	}
	if params.From != "" {
		q.Set("from", params.From)
	}
	if params.To != "" {
		q.Set("to", params.To)
	}
	if params.Limit > 0 {
		q.Set("limit", strconv.Itoa(params.Limit))
	}
	if params.Cursor != "" {
		q.Set("cursor", params.Cursor)
	}
	var env struct {
		Items      []AuditLog `json:"items"`
		Data       []AuditLog `json:"data"`
		NextCursor string     `json:"next_cursor"`
	}
	if err := r.http.do(ctx, http.MethodGet, "/v1/tenants/"+url.PathEscape(tenantID)+"/audit", q, nil, &env, true); err != nil {
		return nil, err
	}
	items := env.Items
	if items == nil {
		items = env.Data
	}
	return &AuditLogPage{Data: items, NextCursor: env.NextCursor}, nil
}

func (r *AuditLogs) Verify(ctx context.Context, tenantID, entryID string) (bool, error) {
	var res struct {
		Valid bool `json:"valid"`
	}
	err := r.http.do(ctx, http.MethodPost, "/v1/tenants/"+url.PathEscape(tenantID)+"/audit/"+url.PathEscape(entryID)+"/verify", nil, struct{}{}, &res, false)
	return res.Valid, err
}

File: sdk/js/node/src/auditLogs.ts
import type { HttpClient } from "./client.js";

export interface AuditLog {
  id: string;
  tenant_id: string;
  actor_id?: string;
  actor_type?: string;
  event: string;
  resource_type?: string;
  resource_id?: string;
  ip_address?: string;
  user_agent?: string;
  metadata?: Record<string, unknown>;
  hash?: string;
  created_at: string;
}

export interface AuditLogListParams {
  tenant?: string;
  event?: string;
  actor_id?: string;
  from?: string;
  to?: string;
  limit?: number;
  cursor?: string;
}

export interface AuditLogPage {
  data: AuditLog[];
  nextCursor?: string;
}

interface ListEnvelope<T> {
  items?: T[];
  data?: T[];
  next_cursor?: string;
}

export class AuditLogs {
  constructor(private readonly http: HttpClient) {}

  async list(tenantId: string, params: AuditLogListParams = {}): Promise<AuditLogPage> {
    const res = await this.http.get<ListEnvelope<AuditLog>>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/audit`,
      {
        query: {
          event: params.event,
          actor_id: params.actor_id,
          from: params.from,
          to: params.to,
          limit: params.limit,
          cursor: params.cursor,
        },
      },
    );
    return { data: res.items ?? res.data ?? [], nextCursor: res.next_cursor };
  }

  async *listAll(tenantId: string, params: AuditLogListParams = {}): AsyncGenerator<AuditLog> {
    let cursor = params.cursor;
    do {
      const page = await this.list(tenantId, { ...params, cursor });
      for (const entry of page.data) yield entry;
      cursor = page.nextCursor;
    } while (cursor);
  }

  verify(tenantId: string, entryId: string): Promise<{ valid: boolean }> {
    return this.http.post<{ valid: boolean }>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/audit/${encodeURIComponent(entryId)}/verify`,
      {},
    );
  }
}
